---
title: "Generating biased gene lists : pre-processing"
author: "Laura Biggins"
date: "4 March 2019"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_knit$set(global.par = TRUE)
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

## Pre-processing of data on cluster
### Generating a gene info file for the mouse genome

The script create_gene_info_file_from_gtf.pl takes a gtf file and parses it to
create a gene info file that contains all the genes that are annotated in the 
gtf file. This should be all the genes in that version of the genome.

Downloaded the raw version of the file from github
`wget https://raw.githubusercontent.com/s-andrews/GOliath/master/processing/`
`gene_info_processing/create_gene_info_file_from_gtf.pl`

Run this script to create the gene info file
`perl create_gene_info_file_from_gtf.pl --gtf Mus_musculus.GRCm38.94.gtf.gz --genome GRCm38`

We'll import the gene info file so that we can plot the GC distribution for all 
genes in the Mus_musculus.GRCm38.94.gtf.gz genome.
There are import_GTF and parse_GTF_info functions within the GOcategoryStats 
package but to get genome information i.e. GC content, the parsing and lookups 
need to be done with access to genome information, so on the cluster. 
The import_GTF and parse_GTF_info functions just work with the gtf file itself.

Import the processed gene info file

```{r}
genfo <- read.delim("M:/biased_gene_lists/Mus_musculus.GRCm38.94_gene_info.txt")
head(genfo)
colnames(genfo)
```

Using this gene info file and some extra processing, lists of genes were generated
for the following categories:
    
1. Length biased gene sets
2. High transcript biased gene sets
3. GC biased gene sets
4. Chromosomal biased gene sets
5. Closest genes to random positions
6. Public data gene sets

The processing for categories 1-4 was carried out within an R session and is detailed 
in the Rmarkdown documents of the same names.

To generate the gene sets for Category 5 - Closest genes to random positions, 
a python script was written. This generated random locations in the genome and found
the closest gene to each position.

Category 6 - the public data required a separate, more extensive workflow.

The processing of each of the 6 categories is detailed in the individual Rmarkdown documents.




