---
title: "5.Closest genes to random positions"
author: "Laura Biggins"
date: "24 January 2019"
output:
  pdf_document: default
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_knit$set(global.par = TRUE)
knitr::opts_chunk$set(echo = FALSE, warning = FALSE)
```

```{r, include = FALSE}
# my laptop doesn't really like working on the network from home, it's better to 
# copy the files to the computer and access them from there. This should make it 
# easier to switch between
using_laptop <- FALSE
```

```{r}
library("devtools")
ifelse(using_laptop,
       load_all("C:/Users/bigginsl/Desktop/temp/GOcategoryStats"),
       load_all("M:/GOcategoryStats")
       )
```

# Closest gene


Some analyses performed on sequencing data use genes as the basis for analysis/ are gene-centric. In RNA-seq for example, a table of count data is often produced where a value for a gene is calculated by counting reads that fall within or overlap the gene. 
Some analyses are not directly based around genes. A common method used in ChIP-seq analysis is peak calling, where a test sample is generally compared to a background sample to identify regions that have a high read count in the test sample. These peaks may be exonic or intronic and can simply be annotated with the overlapping gene. The peaks may also be intergenic in which case they are often annotated with the closest gene.        


Genes vary greatly in size and the spread of genes is not uniform across chromosomes. It may be expected that longer genes, genes in the middle of gene `deserts` and those at the ends of clusters of genes are identified more frequently when annotating data with the closest gene to a region of interest.          


The aim of this study was to see whether any biases were found when annotating random positions within the genome with the closest gene.
This was carried out using the following steps:  


* generating random locations within the mouse genome and annotating these with the closest gene
* seeing whether any particular genes were overrepresented
* performing GO analyses on the lists of closest genes
* investigating the GO categories that were identified multiple times



## Generating genes by random position
  
The script `generate_genelist_from_random_positions.py` produces a list of genes.
It generates a number of random positions along each chromosome. The number of positions generated is proportional to the length of the chromosome. The chromosome lengths are taken from a separate file (`chr_list.txt`)
A gtf file containing all the genes in the genome is required. Each line from the gtf file is parsed and all the genes are stored by chromosome.
Each random position is looped through to find the closest (or overlapping) gene and the output is written to a new file.
This script has been run to create lists of genes for the mouse genome using the gtf file for genome version GRCm38.95

`for i in {1..200}; do qsub  -cwd -V -l h_vmem=10G python ./generate_genelist_from_random_positions.py --output gene_lists_GRCm38.95/closest_genes_${i}.txt; done`

Extract the gene names and remove header
`for i in closest*; do cut -f2 $i | tail -n 200 > ${i%".txt"}_just_genes.txt; done`


```{r cars}
#files <- list.files(path = "M:/biased_gene_lists", pattern = ".txt", full.names = TRUE)
#files <- list.files(path = //fs-home/bigginsl/

closest <- lapply(files, function(x) scan(x, what = "character"))

x <- ifelse(using_laptop,
       genfo <- read.delim("C:/Users/bigginsl/Desktop/temp//biased_gene_lists/Mus_musculus.GRCm38.94_gene_info.txt"),
       genfo <- read.delim("M:/biased_gene_lists/Mus_musculus.GRCm38.94_gene_info.txt")
)

bg_genes <- as.vector(unique(genfo$gene_name))
```

We're going to run the overrepresentation test with no multiple testing correction,
as we want to see which categories come up the most frequently, rather than whether 
they are highly significant.
```{r overrep, eval = FALSE}
go_results1 <- overrep_test(all_go_categories, closest[[1]], bg_genes)
go_results1
go_results2 <- overrep_test(all_go_categories, closest[[1]], bg_genes)
go_results2

all_go_results <- lapply(closest, function(query){
  overrep_test(all_go_categories, query, bg_genes, pval_threshold = 0.2)
  })

#save(all_go_results, file = "D:/projects/biases/GC_content_biased/lowGC/lowGC_results.rda")
save(all_go_results, file = "M:/GOcategoryStats/data/closest_gene.rda")
# this only took a minute or 2 but still worth saving
```

```{r }
ifelse(using_laptop,
       load("C:/Users/bigginsl/Desktop/temp//GOcategoryStats/data/closest_gene.rda"),
       load("M:/GOcategoryStats/data/closest_gene.rda")
)

all_go_results[[6]]

x <- sapply(all_go_results, is.null)

significant_results <- all_go_results[which(!x)]

all_sig_categories <- unlist(sapply(significant_results, rownames))
tabled_categories  <- table(all_sig_categories)
ordered_categories <- tabled_categories[order(tabled_categories, decreasing = TRUE)]
```


