<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Laura Biggins" />


<title>GO_category_summary</title>

<script src="7.GO_category_summary_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="7.GO_category_summary_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="7.GO_category_summary_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="7.GO_category_summary_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="7.GO_category_summary_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="7.GO_category_summary_files/navigation-1.1/tabsets.js"></script>
<link href="7.GO_category_summary_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="7.GO_category_summary_files/highlightjs-9.12.0/highlight.js"></script>
<link href="7.GO_category_summary_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="7.GO_category_summary_files/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">GO_category_summary</h1>
<h4 class="author">Laura Biggins</h4>
<h4 class="date">22/08/2019</h4>

</div>


<pre class="r"><code>library(&quot;devtools&quot;)
load_all()
library(tidyverse)
library(DT)
library(RColorBrewer)
library(gridExtra)
library(formattable)</code></pre>
<div id="go-category-summary" class="section level2">
<h2>GO category summary</h2>
<p>This analysis is looking at the GO categories that have been identified using the biased methods of generating gene lists. I’d like some summary statistics. 1. For each GO category in the file that was used to</p>
<p>where was the list of genes obtained from? I think this was a gtf file of all the genes in the genome, but how many of these were in the GO categories? How many annotations does each gene have? This is getting in to territory that has been explored elsewhere.</p>
<p>I think we want to establish which of our results can identify biases in the data rather than just general rubbish.</p>
<p>I generated completely random lists of genes and ran these through the same gene ontology analysis. Two gene sets out of 200 returned any results at all.</p>
<p>Could it be that the genes that have interesting properties, like length, location, high number of transcripts stand out from the rest so get studied more, get annotated more and therefore generate more significant GO results?</p>
<p>Hopefully this is the case for some, and then for some, we can actually use these categories to identify biases.</p>
<pre class="r"><code>#suspect_cats &lt;- read_tsv(&quot;../data/suspect_categories.txt&quot;)
suspect_cats &lt;- read_tsv(&quot;../data/suspect_categories_archive.txt&quot;) # I think this was the 30 limit
suspect_cats %&gt;%
  count(bias_source) %&gt;%
  arrange(desc(n))</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["bias_source"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"high_transcripts","2":"1206"},{"1":"long","2":"996"},{"1":"public_data","2":"489"},{"1":"vv_high_transcripts","2":"466"},{"1":"very_long","2":"274"},{"1":"very_high_transcripts","2":"270"},{"1":"closest_genes","2":"134"},{"1":"chrMT","2":"132"},{"1":"chr19","2":"128"},{"1":"chr12","2":"69"},{"1":"chr17","2":"31"},{"1":"very_short","2":"25"},{"1":"low_GC","2":"24"},{"1":"chr11","2":"21"},{"1":"chr18","2":"19"},{"1":"chr2","2":"16"},{"1":"chrY","2":"11"},{"1":"very_low_GC","2":"8"},{"1":"short","2":"7"},{"1":"vv_low_GC","2":"6"},{"1":"chr7","2":"4"},{"1":"chr13","2":"3"},{"1":"chr6","2":"2"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>suspect_cats %&gt;%
  count(GO_ID) %&gt;%
  count(n)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["n"],"name":[1],"type":["int"],"align":["right"]},{"label":["nn"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"1","2":"1075"},{"1":"2","2":"719"},{"1":"3","2":"328"},{"1":"4","2":"117"},{"1":"5","2":"43"},{"1":"6","2":"15"},{"1":"7","2":"9"},{"1":"8","2":"1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>get_IDs_in_category &lt;- function(category){
  
    filter(suspect_cats, bias_source == category) %&gt;%
      pull(GO_ID)
}  
all_bias_categories &lt;- suspect_cats %&gt;%
  distinct(bias_source) %&gt;%
  pull(bias_source)
get_count_for_cat &lt;- function(category){
  suspect_cats %&gt;%
  filter(GO_ID %in% get_IDs_in_category(category)) %&gt;%
  count(GO_ID) %&gt;%
  count(n)
}</code></pre>
<p>Add a bias_type column that is almost the same as bias_source, but groups some of the categories together so that if some categories are found in 2 datasets - eg. long and high transcripts, and long and very high transcripts, and long and vv high transcripts, this is just shown as long and high transcripts. This makes some of the plots clearer.</p>
<pre class="r"><code>suspect_cats &lt;- suspect_cats %&gt;%
  mutate(bias_type = if_else(grepl(&quot;chr&quot;, bias_source), &quot;chr&quot;, bias_source)) %&gt;%
  mutate(bias_type = str_replace_all(bias_type, &quot;very_high_transcripts&quot;, &quot;high_transcripts&quot;)) %&gt;%
  mutate(bias_type = str_replace_all(bias_type, &quot;vv_high_transcripts&quot;, &quot;high_transcripts&quot;)) %&gt;%
  mutate(bias_type = str_replace_all(bias_type, &quot;very_long&quot;, &quot;long&quot;))</code></pre>
<pre class="r"><code>plot_more_detail &lt;- function(category){
  suspect_cats %&gt;%
    filter(GO_ID %in% get_IDs_in_category(category)) %&gt;%
    count(GO_ID) %&gt;%
    left_join(suspect_cats) %&gt;%
    arrange(desc(n)) %&gt;%
    group_by(n) %&gt;%
    count(bias_type) %&gt;%
    ggplot(aes(x = n, y = nn, fill = bias_type)) + 
      geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;) +
      ggtitle(category) +
      xlab(&quot;Number of bias sources GO category appeared in&quot;) + 
      ylab(&quot;Number of GO categories&quot;) 
}
p &lt;- lapply(all_bias_categories, plot_more_detail)
names(p) &lt;- all_bias_categories
grid.arrange(p[[&quot;vv_high_transcripts&quot;]], p[[&quot;very_high_transcripts&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-4-1.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;high_transcripts&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-4-2.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;very_long&quot;]], p[[&quot;long&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-4-3.png" width="1152" /></p>
<p>The vv high transcripts plot shows that 357 GO categories are identified only from the genes with a very high number of transcripts. Just under 100 are found in both high transcripts and long gene sets.</p>
<p>Most (159) of the categories identified from the very_high_transcripts set of genes are also only identified from that single category.</p>
<p>For the high transcripts there are still 246 GO categories only identified from that single category, though more 30 are identified in 2 categories, the 2nd category being long genes.</p>
<p>The very_long set identified 96 GO categories unique to the very long set, though more (96) are found in both high transcripts and long sets</p>
<pre class="r"><code>grid.arrange(p[[&quot;very_short&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-6-1.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;low_GC&quot;]], p[[&quot;very_low_GC&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-6-2.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;public_data&quot;]], p[[&quot;closest_genes&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-7-1.png" width="1152" /></p>
<p>The closest gene set does not have any GO categories that are unique to the closest gene set. They all also appear in long, high_transcripts, and most appear in the public data sets as well. This seems to indicate that these categories find significance too easily. But, they didn’t come up when completely random sets of genes were used, so it may be that these properties of genes have made them more likely to be investigated and annotated??<br />
We’ll try and look into this by seeing how many GO annotations these sets of genes have.</p>
<p>The GC and chr plots make more sense when the categories are divided in to separate chromosomes.</p>
<pre class="r"><code>plot_more_detail &lt;- function(category){
  suspect_cats %&gt;%
    filter(GO_ID %in% get_IDs_in_category(category)) %&gt;%
    count(GO_ID) %&gt;%
    left_join(suspect_cats) %&gt;%
    arrange(desc(n)) %&gt;%
    group_by(n) %&gt;%
    count(bias_source) %&gt;%
    ggplot(aes(x = n, y = nn, fill = bias_source)) + 
      geom_bar(stat = &quot;identity&quot;, position = &quot;dodge&quot;) +
      ggtitle(category) +
      xlab(&quot;Number of bias sources GO category appeared in&quot;) + 
      ylab(&quot;Number of GO categories&quot;) 
}
p &lt;- lapply(all_bias_categories, plot_more_detail)
names(p) &lt;- all_bias_categories
grid.arrange(p[[&quot;low_GC&quot;]], p[[&quot;very_low_GC&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-8-1.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;chr11&quot;]], p[[&quot;chr12&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-8-2.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;chr17&quot;]], p[[&quot;chr18&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-8-3.png" width="1152" /></p>
<pre class="r"><code>grid.arrange(p[[&quot;chr19&quot;]], p[[&quot;chrMT&quot;]], ncol = 2)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-8-4.png" width="1152" /></p>
</div>
<div id="gmt-file" class="section level2">
<h2>GMT file</h2>
<pre class="r"><code># load the set of genes and GO categories - this is probably a very inefficient way of holding the data but it can be changed if it causes problems
load(&quot;../data/gathered_cats.rda&quot;)
categories_per_gene &lt;- count(gathered_cats, gene)



load(&quot;../data/closest_gene.rda&quot;)
all_genes &lt;- unlist(closest_gene_lists)
tabled_genes &lt;- table(all_genes)
tabled_genes &lt;- tabled_genes[order(tabled_genes, decreasing = TRUE)]
gene_counts  &lt;- data.frame(gene_name = names(tabled_genes), n = as.vector(tabled_genes))</code></pre>
<pre class="r"><code># load the set of genes and GO categories - this is probably a very inefficient way of holding the data but it can be changed if it causes problems
load(&quot;../data/gathered_cats.rda&quot;)
load(&quot;../data/closest_gene.rda&quot;)
all_genes &lt;- unlist(closest_gene_lists)
tabled_genes &lt;- table(all_genes)
tabled_genes &lt;- tabled_genes[order(tabled_genes, decreasing = TRUE)]

gene_counts  &lt;- data.frame(gene_name = names(tabled_genes), n_closest = as.vector(tabled_genes))

categories_per_gene &lt;- count(gathered_cats, gene)

closest_gene_go_counts &lt;- gene_counts %&gt;%
  mutate(gene_name = toupper(gene_name)) %&gt;%
  full_join(categories_per_gene, by = c(&quot;gene_name&quot; = &quot;gene&quot;)) %&gt;%
  replace_na(list(n_closest = 0)) %&gt;%
  arrange(desc(n))

go_annot_summary &lt;- plyr::rename(closest_gene_go_counts, c(&quot;n&quot; = &quot;n_GO_cats&quot;))</code></pre>
<pre class="r"><code># load the set of genes and GO categories - this is probably a very inefficient way of holding the data but it can be changed if it causes problems
load(&quot;../data/top_200.rda&quot;) # the public data
all_genes &lt;- unlist(top200)
go_annot_summary &lt;- count(as.data.frame(all_genes), all_genes) %&gt;%
  plyr::rename(c(&quot;all_genes&quot; = &quot;gene_name&quot;, &quot;n&quot; = &quot;n_public&quot;)) %&gt;%
  mutate(gene_name = toupper(gene_name)) %&gt;%
  full_join(go_annot_summary) %&gt;%
  replace_na(list(n_public = 0))  %&gt;%
  arrange(desc(n_public))

# the top couple of genes from the public set are not in the gmt file

load(&quot;../data/biased_transcripts.rda&quot;)
process_transcript_genes &lt;- function(list_name){
  all_genes &lt;- as.vector(unlist(biased_transcripts[[list_name]]))
  column_name &lt;- paste0(&quot;n_&quot;, list_name)
  count(as.data.frame(all_genes), all_genes) %&gt;%
  plyr::rename(c(&quot;all_genes&quot; = &quot;gene_name&quot;, &quot;n&quot; = column_name)) %&gt;%
  mutate(gene_name = toupper(gene_name)) %&gt;%
  right_join(go_annot_summary)
}
  
go_annot_summary &lt;- process_transcript_genes(&quot;0.9&quot;)
go_annot_summary &lt;- process_transcript_genes(&quot;0.95&quot;)
go_annot_summary &lt;- process_transcript_genes(&quot;0.99&quot;)

go_annot_summary &lt;- go_annot_summary %&gt;%
   replace(is.na(.), 0) %&gt;%
   mutate(transcript_total = rowSums(.[2:4]))

go_annot_summary &lt;- go_annot_summary %&gt;%
  mutate(min_per_row = apply(go_annot_summary[,c(5,6,8)], 1, min))
 
go_annot_summary &lt;- go_annot_summary %&gt;%
  mutate(median_per_row = apply(go_annot_summary[,c(5,6,8)], 1, median)) 

plot(go_annot_summary$min_per_row, go_annot_summary$n_GO_cats)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<pre class="r"><code>plot(go_annot_summary$median_per_row, go_annot_summary$n_GO_cats)</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-12-2.png" width="672" /></p>
<pre class="r"><code>plot(density(go_annot_summary$n_GO_cats))</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-12-3.png" width="672" /></p>
<pre class="r"><code>go_annot_no_0 &lt;- filter(go_annot_summary, n_GO_cats &gt; 0)
plot(density(go_annot_no_0$n_GO_cats))</code></pre>
<p><img src="7.GO_category_summary_files/figure-html/unnamed-chunk-12-4.png" width="672" /></p>
<pre class="r"><code>#boxplot(log2((go_annot_no_0$n_GO_cats)))</code></pre>
<p>Not really sure what to do with this…</p>
<p>Select some genes to make an example gene list for GOliath</p>
<pre class="r"><code>g1 &lt;- filter(go_annot_summary, transcript_total &lt;2 &amp; n_public &lt; 2 &amp; n_closest &lt; 2 &amp; n_GO_cats &gt; 1) %&gt;%
  pull(gene_name)



eg_set_of_genes &lt;- g1[ceiling(runif(200, min = 0, max = length(g1) - 1))]</code></pre>
<pre class="r"><code>x &lt;- c(&quot;GM14483&quot;, &quot;SSXB10&quot;, &quot;FAM87B&quot;, &quot;MINOS1-NBL1&quot;, &quot;SMAP2&quot;,&quot;CTPS1&quot;, &quot;MOB3C&quot;, &quot;DLEU2L&quot;, &quot;KIAA1324&quot;, &quot;CD58&quot;,&quot;ZNF697&quot;, &quot;IL6R&quot;, &quot;SLAMF9&quot;, &quot;SH2D1B&quot;, &quot;FASLG&quot;,&quot;LINC01686&quot;, &quot;PACERR&quot;, &quot;FAM89A&quot;, &quot;GCSAML&quot;, &quot;LUCAT1&quot;,&quot;EPB41L4A&quot;, &quot;GRAMD2B&quot;, &quot;TXNDC15&quot;, &quot;LINC01024&quot;, &quot;SMIM3&quot;,&quot;LINC01366&quot;, &quot;PXDC1&quot;, &quot;HCG18&quot;, &quot;TRIM39-RPP21&quot;, &quot;MICA&quot;,&quot;HCP5&quot;, &quot;MICB&quot;)</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
